contracts/bulk-payments/src/bulk_payments.rs

```diff
@@ -0,0 +1,140 @@
+        // Safely unwrap cause: `number_users_opted_in >= MAX_USERS_ALLOW``
+        let first_user = self.opted_in_addrs().front().unwrap();
+
+        require!(
+            self.is_timestamp_expired(first_user.clone()),
+            "Max CAP reached"
+        );
+        self.addr_timestamp(first_user.clone()).clear();
+        self.opted_in_addrs().remove(&first_user);
+    }
+
+    #[only_owner]
+    #[endpoint(whitelistUsers)]
+    fn add_whitelist_users(&self, users: MultiValueEncoded<ManagedAddress>) {
+        let whitelist_mapper = self.user_whitelist();
+        for user in users {
+            whitelist_mapper.add(&user);
+        }
+    }
+
+    #[only_owner]
+    #[endpoint(removeWhitelistWsers)]
+    fn remove_whitelist_users(&self, users: MultiValueEncoded<ManagedAddress>) {
+        let whitelist_mapper = self.user_whitelist();
+        for user in users {
+            whitelist_mapper.remove(&user);
+        }
+    }
+
+    /// # views
+    #[view(isUserWhitelisted)]
+    fn is_user_whitelisted(&self, user: ManagedAddress) -> bool {
+        self.user_whitelist().contains(&user)
+    }
+
+    #[view(isUserOptedIn)]
+    fn is_user_opted_in(&self, user: ManagedAddress) -> bool {
+        return self.opted_in_addrs().contains(&user);
+    }
+
+    #[view(isTimestampExpired)]
+    fn is_timestamp_expired(&self, user: ManagedAddress) -> bool {
+        let user_timestamp = self.addr_timestamp(user).get();
+
+        user_timestamp < self.blockchain().get_block_timestamp()
+    }
+
+    // Returns user without timestamp expired
+    #[view(getUsersOptedIn)]
+    fn get_users_opted_in(&self) -> MultiValueEncoded<ManagedAddress> {
+        let mut users_opted_in = MultiValueEncoded::new();
+        let addr_timestamp_mapper = self.opted_in_addrs();
+        let current_timestamp = self.blockchain().get_block_timestamp();
+
+        for user_addr in addr_timestamp_mapper.iter() {
+            let timestamp = self.addr_timestamp(user_addr.clone()).get();
+            if timestamp > current_timestamp {
+                users_opted_in.push(user_addr);
+            }
+        }
+        return users_opted_in;
+    }
+
+    // Total user without timestamp expired
+    #[view(getEligibleNumberUsersOptedIn)]
+    fn get_eligible_number_users_opted_in(&self) -> usize {
+        return self.get_users_opted_in().len();
+    }
+
+    // Total user with timestamp expired and not expired
+    #[view(getNumberUsersOptedIn)]
+    fn get_number_users_opted_in(&self) -> usize {
+        return self.opted_in_addrs().len();
+    }
+
+    #[view(getUserTimestamp)]
+    fn get_user_timestamp(&self, user: ManagedAddress) -> u64 {
+        return self.addr_timestamp(user).get();
+    }
+
+    #[storage_mapper("addrTimestamp")]
+    fn addr_timestamp(&self, user: ManagedAddress) -> SingleValueMapper<u64>;
+
+    #[storage_mapper("optedInAddrs")]
+    fn opted_in_addrs(&self) -> SetMapper<ManagedAddress<Self::Api>>;
```

Comment:
Being a queue basically, I would recommend to use LinkedListMapper, which is made mostly for these types of cases. It allows you to better add/remove from the front / back of the list and it is more efficient.

Reply:
This type of that has high cost of using `contains` (which we use often).